add heap hint support for a few of the x509 functions

jpbland1 opened the pull request:
# Description

Adds heap hint support only for the functions the customer asked for

Fixes zd# 17258

# Testing

Uses api and crl tests

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

dgarske commented on src/x509_str.c:
```diff
@@ -1035,7 +1035,7 @@ WOLFSSL_API int wolfSSL_X509_STORE_load_locations(WOLFSSL_X509_STORE *str,
 
     /* tmp ctx for setting our cert manager */
-    ctx = wolfSSL_CTX_new(cm_pick_method());
+    ctx = wolfSSL_CTX_new(cm_pick_method(NULL));
```
Use `str->cm->heap`

jpbland1 replied:
ah ok, str->cm is checkedfor NULL

dgarske commented on wolfssl/ssl.h:
```diff
@@ -1681,7 +1681,9 @@ WOLFSSL_API void wolfSSL_sk_CIPHER_free(WOLF_STACK_OF(WOLFSSL_CIPHER)* sk);
 WOLFSSL_API WOLFSSL_X509* wolfSSL_X509_new(void);
+WOLFSSL_API WOLFSSL_X509* wolfSSL_X509_new_ex(void* heap);
 WOLFSSL_API WOLFSSL_X509* wolfSSL_X509_dup(WOLFSSL_X509* x);
+WOLFSSL_API WOLFSSL_X509* wolfSSL_X509_dup_ex(WOLFSSL_X509* x, void* heap);
```
Do you really need `wolfSSL_X509_dup_ex`? Can't the `wolfSSL_X509_dup` just use `x->heap` from the previous one? If you have places where heap isn't populated then switch those to using `wolfSSL_X509_new_ex`.

jpbland1 replied:
you're right, I had this in my other PR but was following what the customer gave me for this one

jpbland1 commented:
retest this please

jpbland1 commented:
I'm getting:
```
Got -132 but expected -171 at line 4947 in ecdsa_invalid_data_tests
Encountered an unexpected result
 ECDSA-INVALID ... FAILED
 ```
which I don' think I've changed, lets try one more time:
retest this please

dgarske commented:
@jpbland1 . This is a known issue. @lealem47 is working on it.

dgarske reviewed (approved):
Looks good. Thanks

lealem47 commented:
Jenkins Retest this please

dgarske reviewed (changes requested):
Please add a test case for `wolfSSL_X509_d2i_ex` or adapt an existing one. How have you confirmed all X509 nodes have the heap hint? Did you try with `WOLFSSL_HEAP_TEST`?

jpbland1 commented:
> Please add a test case for `wolfSSL_X509_d2i_ex` or adapt an existing one. How have you confirmed all X509 nodes have the heap hint? Did you try with `WOLFSSL_HEAP_TEST`?

Updated the test cases but the question of how to make sure all X509 nodes have the heap hint is why I didn't want to do a half measure on this but we can't blow up the x509small size. I've updated the functions that already have the heap hint and wouldn't require their own ex function and I guess if a customer needs those functions too we can add them but I think we need a long term solution so x509small doesn't break.

Tested working with `WOLFSSL_HEAP_TEST`.

Comparing the sizes, configured with `./configure --enable-opensslextra=x509small`
On 089468fbf16652161e0ef53e54ccf47e3db24104:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
    445915     198363      24200     668478 /usr/local/lib/libwolfssl.so.42.0.0
```
On current commit:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
    445979     198535      24200     668714 /usr/local/lib/libwolfssl.so.42.0.0
```
A difference of 236 bytes

JacobBarthelmeh commented on src/ssl_certman.c:
```diff
@@ -876,7 +876,7 @@ int wolfSSL_CertManagerLoadCA(WOLFSSL_CERT_MANAGER* cm, const char* file,
     }
     /* Create temporary WOLFSSL_CTX. */
-    if ((ret == WOLFSSL_SUCCESS) && ((tmp = wolfSSL_CTX_new(cm_pick_method()))
+    if ((ret == WOLFSSL_SUCCESS) && ((tmp = wolfSSL_CTX_new(cm_pick_method(cm->heap)))
```
This should also be wolfSSL_CTX_new_ex to avoid losing memory assigned to the method pointer on WOLFSSL_CTX free.

JacobBarthelmeh commented on src/x509_str.c:
```diff
@@ -1035,7 +1036,7 @@ WOLFSSL_API int wolfSSL_X509_STORE_load_locations(WOLFSSL_X509_STORE *str,
 
     /* tmp ctx for setting our cert manager */
-    ctx = wolfSSL_CTX_new(cm_pick_method());
+    ctx = wolfSSL_CTX_new(cm_pick_method(str->cm->heap));
```
Use wolfSSL_CTX_new_ex here.
